packages/squiggle-lang/src/public/SqValue/index.ts

```diff
@@ -68,7 +74,7 @@ export function wrapValue(value: Value, context?: SqValueContext) {
   }
 }
 
-export abstract class SqAbstractValue<Type extends string, JSType> {
+export abstract class SqAbstractValue<Type extends string, JSType, ValueType> {
```

Comment:
This extra type seemed safe to have. I think I needed it for ``getSubvalueByPath`` at some point - not sure if I need it now, but it seems good to have anyway.

Reply:
If you use it only for `value()`, I don't think it's useful.

If you had a value that's proven to be a specific `SqValue` (e.g. `SqNumberValue`), you could extract this:

<img width="591" alt="Screenshot 2024-01-23 at 15 20 00" src="https://github.com/quantified-uncertainty/squiggle/assets/89368/0137fe2c-0bb3-4c18-a266-46ce1d195a9f">

OTOH, if you add a type parameter like this, then you have to provide it in every subclass, when previously TypeScript would happily infer it for you.

(There's still [no partial type argument inference in TypeScript](https://github.com/microsoft/TypeScript/issues/26242))

---

There's a longer sub-discussion here about when we should explicitly provide types that could be inferred.

I've noticed that you prefer to write out types more often than I do.

I generally like to rely on type inference, except for cases when I want to guarantee exhaustiveness or have a function with multiple branches.

Example:
```typescript
function convertToDist(x: Foo) {
  if (x.type === "a") {
    return x.asDist();
  } else if (x.type === "b") {
    return processXOfBType(x);
  }
}
```

In this case, I'd be tempted to type `convertToDist(x: Foo): Dist` explicitly, if either of these is true:
- I don't want to write an extra `else { throw new Error(``Unknown x type ${x.type satisifies never}`` }`, and want to guarantee that this code won't break when `x` gets another possible type
- I'm not confident that `processXOfBType` returns a dist (though then I'd prefer to make that function more reliable and leave this function as is)

In most other cases, I think implicit type inference is a better choice, especially if the return type is complex, which it tends to be, sometimes.

When you have to write something like `: result<Dist, SqError>` on all functions that pass that result around, that's basically copy-paste, and you would have to change the code in many more locations when you need to edit it.

Btw, if you prefer to write explicit types because hovering over values is annoying, there's a recent [inlay hints](https://code.visualstudio.com/docs/typescript/typescript-editing#_inlay-hints) option in VS Code that can show inferred types.

---

If we do write out types, I also would prefer to do it close to the location that creates the value (functions as a source of truth). With this generic parameter, we have to write `ValueType` at the first line of class definition, but the source of truth is `get value()`, which is declared separately.

Btw, my opinion on the value of this would change if you used it in multiple places. If we had not just `get value(): ValueType`, but also `fromValue(v: ValueType)`, then sure, let's do it.

---

One last point: I realize that I don't remember why we need separate `SqCalculatorValue` vs `SqCalculator`, etc. They're parameterized by the same data, maybe we can simplify it.